core/rawdb/freezer_table.go

```diff
@@ -403,6 +375,106 @@ func (t *freezerTable) truncate(items uint64) error {
 	return nil
 }
 
+func (t *freezerTable) fileName(number int) string {
+	if t.noCompression {
+		return fmt.Sprintf("%s.%04d.rdat", t.name, number)
+	}
+	return fmt.Sprintf("%s.%04d.cdat", t.name, number)
+}
+
+// tailTruncate discards any old data below the provided threshold number.
+func (t *freezerTable) tailTruncate(threshold uint64) error {
+	t.lock.Lock()
+	defer t.lock.Unlock()
+	var (
+		tail = t.itemOffset
+		head = atomic.LoadUint64(&t.items)
+	)
+	if tail >= threshold {
+		return nil
+	}
+	if threshold >= head {
+		return errors.New("tail truncation above head")
+	}
+	// We need to truncate, save the old size for metrics tracking
+	oldSize, err := t.sizeNolock()
+	if err != nil {
+		return err
+	}
+	log := t.logger.Debug
+	if threshold-tail > 1 {
+		log = t.logger.Warn // Only loud warn if we delete multiple items
+	}
+	log("Tail-truncating freezer table", "from", tail, "to", threshold, "items", threshold-tail)
+	// Determine what file the new first item resides in
+	indexes, err := t.getIndices(threshold, 1)
+	if err != nil {
+		return err
+	}
+	start, _, fileNo := indexes[0].bounds(indexes[1])
+	//Close/release all files up to and including fileNo
+	t.releaseFilesBefore(fileNo, false)
+	// In file corresponding to fileNo, we now need to delete all data prior to 'start'
+	destPath := t.fileName(int(fileNo))
+	copyFrom(destPath, destPath, uint64(start))
```

Comment:
Since we truncate the data file first, a few corner cases need to be considered

e.g. if crash happens before we update the index file 

- how to handle the items which are deleted from the data file but not from the index file yet(non-existent?)
- how to handle the items which are shifted in the tail data file but the relevant indexes are not updated yet(wrong item accessed with the original index?)